File: javascript/bird-watcher/bird-watcher.js
// @ts-check

/**
 * Callback for array method reduce() to get total sum of bird visits.
 * 
 * @param {number} partialSum - bird visits accumulator
 * @param {number} visitCount - current number of bird sights in one day
 * @returns {number} partial sum of bird sights since logs started
 */
const birdVisits = (partialSum, visitCount) => partialSum + visitCount;

/**
 * Calculates the total bird count.
 *
 * @param {ArrayLike<number>} birds - log of daily bird sightings in the garden
 * @returns {number} total bird count since logs started
 */
export const totalBirdCount = birds => Array.from(birds).reduce(birdVisits);

/**
 * Calculates the total number of birds seen in a specific week.
 *
 * @param {ArrayLike<number>} birds - log of daily bird sightings in the garden
 * @param {number} week - a 1-based index representing a chunk of 7 indices
 * @returns {number} birds counted in the given week
 */
export function birdsInWeek(birds, week) {
  return totalBirdCount(Array.from(birds).slice(7 * (week - 1), 7 * week)); }

/**
 * Fixes the untracked hidden bird mistake by increasing the bird count by one
 * in place for every second day (even indices).
 *
 * @param {number[]} birds - log of daily bird sightings in the garden
 */
export function fixBirdCountLog(birds) {
  for (var i = 0; i < birds.length; i += 2) ++birds[i]; }

File: go/bird-watcher/bird_watcher.go
package birdwatcher

// TotalBirdCount return the total bird count by summing
// the individual day's counts.
//
// Param:
//
// birds: a slice of integers where each integer represents
// the number of birds counted on a particular day in the garden.
//
// Returns:
//
// total: an integer representing the total count of birds since logs started.
func TotalBirdCount(birds []int) (total int) {
	for i, len := 0, len(birds); i < len; i++ { total += birds[i] }; return }

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
//
// Params:
//
// birdsPerDay: a slice of integers where each integer represents
// the number of birds counted on a particular day.
//
// week: a 1-based index representing a chunk of 7 indices.
//
// Returns:
//
// An integer representing the total count of birds for the specified week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	week *= 7; return TotalBirdCount( birdsPerDay[week - 7 : week] ) }

// FixBirdCountLog fixes the untracked hidden bird mistake by increasing the
// bird count by one in place for alternate days (even indices).
//
// Param:
//
// birds: a slice of integers representing the log of daily bird sightings
// in the garden.
//
// Returns:
//
// The corrected bird count slice, incremented for every even index.
func FixBirdCountLog(birds []int) []int {
	for i, len := 0, len(birds); i < len; i += 2 { birds[i]++ }; return birds }
